plugin/keyring_vault/vault_memory.h

```diff
@@ -0,0 +1,47 @@
+#ifndef MYSQL_VAULT_MEMORY
+#define MYSQL_VAULT_MEMORY
+
+#include <my_global.h>
+#include <string.h>
+#include "keyring_memory.h"
+
+namespace keyring
+{
+#if !defined(HAVE_STDCXX11) && !defined(WIN32)
+  inline void memset_s(void *dest, size_t dest_max, int c, size_t n)
+  {
+    volatile unsigned char *p = reinterpret_cast<unsigned char*>(dest);
+    while (dest_max-- && n--) {
+      *p++ = c;
+    }
+  } 
+#endif
+
+  template <class T> class Secure_allocator : public std::allocator<T>
```

Comment:
That would be true if a Secure_allocator were passed to std::string by pointer or reference, but it is in fact constructed through compile type template, i.e.
 template < class charT,
           class traits = char_traits<charT>,    // basic_string::traits_type
           class Alloc = allocator<charT>        // basic_string::allocator_type
           > class basic_string;
so, no virtual methods needed. I understand your concerns about misusing this class in a future, however, I find it very unlikely that it would be used outside std containers, where allocator are passed as an object.